### PR TITLE
Update a Boefje

### LINKED ISSUE #3468: Create Boefjes - Edit Boefje
### Detailed description
It should be possible to edit the details/settings of the created Boefjes.

### Feature benefit/User story
As a user, I want to edit my Boefje, so that I can quickly make changes.

### Specifications
The implementation should include…

- [ ] "Edit variant" button
- [ ] Filled in 'Edit "Boefje -name-"' page

### Screenshots

_Include screenshots of the proposed design changes here._

### Figma link
https://www.figma.com/design/D48Pf9lLpxET1jp0Wsjzw5/%E2%9C%85-KAT-%7C-Approved-%7C-Boefjes?node-id=2301-14505&node-type=CANVAS&t=FjlhjSWyTlHzn3Kd-0

## Implementation

_This part should only be filled in by the developers._

### Possible solution

_Outline your proposed solution for implementing the feature. You can include any specific ideas, designs, or functionalities here._

### Alternatives considered

_Describe any alternative approaches or solutions you've considered, and why you believe the proposed solution is superior._

### PR DESCRIPTION
**Pull request by madelondohmen**
**NOTE:** This PR is a follow-up of PR #3456 

------------------------

### Changes
This PR makes it possible to update a Boefje variant.

#### Features added
**Boefje detail page:**
- "Edit variant" button on Boefje detail page (in variants table)
- Variants table: add "Boefje ID"
- Variants table: add age column
- Variants table: fix created date (make it readable)

**New Boefje form page:**
- Change submit/discard button names to "Create new Boefje" and "Discard new Boefje"

**Edit Boefje form page:**
- "Edit Boefje" form page that is pre-filled
- Update helpertext for "Container image" input field
- Return to updated Boefje page

**Boefje variant form page:**
- Fix breadcrumbs
- Remove read-only from "Container image" input field

**Code:**
- In View: improve code and reduce duplicate code

### Issue link
Closes #3468

### Demo
afbeelding
afbeelding

### QA notes
First create a variant of Nmap TCP:
- Go to "Nmap TCP" boefje
- Check the "Variants" table. There should be one item (the current Boefje)
- Click "Add variant"
- Type a different name, make some small changes to the other input fields
- Click "Create variant"
- You should return to the "Nmap TCP" boefje

Then, edit the variant:
- Check the "Variants" table. Unfold your just created variant.
- Click "Edit variant"
- Check if **all** fields in de form are pre-filled
- Type a different name, make some small changes to the other input fields
- Click "Save changes"
- You should return to the variant boefje detail page
- Check if the changes were saved

---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [x] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

- [ ] Tickets have been created for newly discovered issues.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/katalogus/views/boefje_setup.py [hunk 1]
```diff
@@ -31,31 +60,139 @@ def get_context_data(self, **kwargs):
...
+        super().setup(request, *args, **kwargs)
+
+        self.plugin_id = str(uuid.uuid4())
+        self.return_to_plugin_id = self.kwargs.get("plugin_id")
+        katalogus = get_katalogus(self.organization.code)
+        self.plugin = katalogus.get_plugin(self.return_to_plugin_id)
+        consumes = []
+
+        for input_object in list(self.plugin.consumes):
+            consumes.append(input_object.__name__)
```
Comment (ammar92) [inline rocky/katalogus/views/boefje_setup.py:77]: You can iterate over sets like any other sequence, no need to create a list
```suggestion
        for input_object in self.plugin.consumes:
            consumes.append(input_object.__name__)
```

File: rocky/katalogus/views/boefje_setup.py [hunk 2]
```diff
@@ -31,31 +60,139 @@ def get_context_data(self, **kwargs):
...
         return context
 
+
+class AddBoefjeVariantView(BoefjeSetupView):
+    """View where the user can create a Boefje variant, based on another Boefje."""
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+
+        self.plugin_id = str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (madelondohmen) [reply rocky/katalogus/views/boefje_setup.py:70]: It looks like they're the same, but they're slightly different. `plugin_id` is the ID of the Boefje that's been created or edited. `return_to_plugin_id` is the id of the Boefje you'll be returned to after saving. `plugin` is the plugin object that belongs to the ID. 

Let me explain:

### AddBoefjeView
- `plugin_id`: Create new Boefje with new uuid
- `return_to_plugin_id`: Return to this new uuid
- `plugin` is not used here

### AddBoefjeVariantView
- `plugin_id`: Create new Boefje with new uuid
- `return_to_plugin_id`: Return to the Boefje where you've created a variant of
- `plugin`: The plugin object that belongs to plugin_id

### EditBoefjeView
- `plugin_id`: The ID of the Boefje you want to edit
- `return_to_plugin_id`: Return to this Boefje you've editted
- `plugin`: The plugin object that belongs to plugin_id

[the remark this answers is not in the document]
Comment (madelondohmen) [reply rocky/katalogus/views/boefje_setup.py:70]: The variable return_to_plugin_id will be removed in #3573, so this might not be relevant anymore.

**Comment by TwistMeister** [comment]
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
  - [x] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [x] I checked the logs for errors and/or warnings and made issues where necessary

### What works:

- Adding a variant
- Editing a variant

### What doesn't work:
n/a

### Bug or feature?:
n/a

**Review by TwistMeister — APPROVED** [review]

**Review by underdarknl — APPROVED** [review]

**Pull request merged**